### PR TITLE
Store documents in a memdb-backed table

### PR DESCRIPTION
**Pull request by radeksimko**
This is a first part of refactoring per #719 

--- 

### Why

Per LSP contract the language server is basically expected to maintain the up-to-date copy of the document sent by client via text synchronization methods (e.g. `didOpen` or `didChange`). Previously we did this as part of the `filesystem` package where we kept it within a map:

https://github.com/hashicorp/terraform-ls/blob/b13641c6b00755dc78c46b540befa7b2059cf39f/internal/filesystem/filesystem.go#L20-L21

The `filesystem` package also served two use cases, which ended up not being as related to each as we originally thought:
 - unified FS interface to pass e.g. to HCL parser so that it can parse either (virtual) document or the underlying file on the OS FS easily - previously solved largely via `spf13/afero`
 - direct access to the document metadata via read and write methods

While this was a reasonable implementation originally, I believe we have slowly outgrown it with our needs.

Much of the important state which LS maintains is held in memdb and so it makes sense to move this data to memdb as well. We can benefit from the "data locality" as we don't need to be importing another package like `filesystem` if we only need access to documents (which is majority of the RPC layer). Secondly we can query and update data across memdb tables more easily. Lastly it makes the architecture a bit easier to understand when data/state is kept in one place, stored the same way.

### New `document` package

A new package is dedicated to:
 - `Document` type describing the document itself (exactly how it is stored in memdb)
 - Decoupled concept of "handles" which was originally spread across `lsp`, `filesystem` packages and elsewhere. This makes it a single place which deals with conversions between file, document and directory URIs as needed in various places.
 - logic related to calculation of changes between two versions of a document

Having a dedicated (smaller) package makes it easier to import throughout the codebase without running into import cycles.

### New `documents` memdb table

```go
type Document struct {
	Dir      DirHandle
	Filename string

	ModTime    time.Time
	LanguageID string
	Version    int

	// Text contains the document body stored as bytes.
	// It originally comes as string from the client via LSP
	// but bytes are accepted by HCL and io/fs APIs, hence preferred.
	Text []byte

	// Lines contains Text separated into lines to enable byte offset
	// computation for any position-based operations within HCL, such as
	// completion, hover, semantic token based highlighting, etc.
	// and to aid in calculating diff when formatting document.
	// LSP positions contain just line+column but hcl.Pos requires offset.
	Lines source.Lines
}
```

Previously the `filesystem` package maintained a `isOpen` flag next to a document indicating whether the document is open but that flag was never used because we only ever tracked open documents and removed them when they get closed - see https://github.com/hashicorp/terraform-ls/blob/b13641c6b00755dc78c46b540befa7b2059cf39f/internal/filesystem/filesystem.go#L171-L187
It therefore made sense to just avoid that flag.

### Removal of `spf13/afero`

After decoupling documents into memdb and re-plumbing DocumentStore methods back to the `filesystem` FS interface I realized that there's actually relatively little benefit we would now get from the external library and decided to just reimplement the small part that we actually need - which is translation between `document.Document` and `fs.FileInfo` or `fs.DirEntry` which ended in `internal/filesystem/document.go`, `internal/filesystem/inmem.go` and `internal/filesystem/os_fs.go` - altogether relatively few LOC.

Unfortunately we still depend on afero indirectly through `mockery` which we use in tests:
```sh
$ go mod why github.com/spf13/afero
```
```
# github.com/spf13/afero
github.com/hashicorp/terraform-ls/tools
github.com/vektra/mockery/v2
github.com/vektra/mockery/v2/cmd
github.com/spf13/viper
github.com/spf13/afero
```
but this is "dev" dependency only.

### CODE REVIEW

File: internal/state/documents.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+		}
+	}
+	return obj.(*document.Document), nil
+}
+
+func (s *DocumentStore) CloseDocument(dh document.Handle) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	obj, err := txn.First(s.tableName, "id", dh.Dir, dh.Filename)
```
Comment (dbanck) [inline internal/state/documents.go:109]: Would it be possible to use `getDocument` here? The operation looks identical.

**Review by dbanck — APPROVED** [review]
Fantastic work and presentation! Thank you for taking the time to make the review as easy as possible.

I was able to follow through with all of your changes, and the language server still works ;)

I've added a few questions to understand memdb handling transactions better.

File: internal/state/documents.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	tableName string
+	logger    *log.Logger
+
+	// TimeProvider provides current time (for mocking time.Now in tests)
+	TimeProvider func() time.Time
+}
+
+func (s *DocumentStore) OpenDocument(dh document.Handle, langId string, version int, text []byte) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/state/documents.go:23]: We basically just rely on `Abort()` being no-op after successful `Commit()`:
https://github.com/hashicorp/go-memdb/blob/542a5804dedaaddf1a0cad8b6a7b0b3fcf438776/txn.go#L112-L115

This in combination with `defer` means we don't have to duplicate `Abort` in every `return` branch - i.e. we don't end up inadvertently introducing a bug.

We have to _either_ `Commit` or `Abort` any write transaction, otherwise the (internal) write mutex lock remains locked, effectively causing a deadlock and preventing any other future write transactions to hang.

It is true that this is not clear from the relevant upstream comment/docs though - I raised a PR to address that: https://github.com/hashicorp/go-memdb/pull/120

File: internal/state/documents.go (hunk 1 above)
Comment (radeksimko) [reply internal/state/documents.go:109]: It is _almost_ identical. The main use case for `getDocument` is to get the contents, not check existence (which is what we need to do in `CloseDocument`).

It's probably negligible benefit in the context of a single document but it might make some difference at scale (e.g. should the user close many documents in quick succession) as we avoid the type casting in `CloseDocument` from `interface{}` to `*document.Document`.

File: internal/state/documents.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	if obj == nil {
+		// already removed
+		return &document.DocumentNotFound{
+			URI: dh.FullURI(),
+		}
+	}
+
+	_, err = txn.DeleteAll(s.tableName, "id", dh.Dir, dh.Filename)
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/state/documents.go:121]: This was merely a choice of convenience on the memdb API level as we'd have to reconstruct the `document.Document` object, or build the relevant part of it before passing it to `Delete(table string, obj interface{})`.

Technically this is just a type cast away as we check for existence anyway though - so we _could_ use `Delete` here I guess.

I don't know what's better though without also running some more benchmarks and weighing these against the convenience arguments. It is possible that `Delete` would perform better than `DeleteAll` in principle but we look up a single entry based on a unique index, so it probably won't make a difference.

File: internal/document/handle.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package document
+
+import (
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+type Handle struct {
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/document/handle.go:15]: I have attached some tests to the `document` package, but more importantly decided to refactor the `uri` package and move all the important logic there and add more tests there.

**Comment by radeksimko** [comment]
The refactored `uri` package now accounts for the edge case we discovered with Windows-style paths `C:\` & VS Code as per https://github.com/microsoft/vscode/issues/75027 and contains some tests for that as well.

I also moved some of the existing logic we had in place for case-insensitive comparison of URIs there and added some comments after researching the problem.

There is a few problems which this PR does _not_ address, but we have issues to track them separately:
 - https://github.com/hashicorp/terraform-ls/issues/74
 - https://github.com/hashicorp/terraform-ls/issues/656
 - https://github.com/hashicorp/terraform-ls/issues/796

@dbanck Do you mind reviewing the newly added commits? I intentionally left the old ones intact, so it should be easier to filter out.

I re-tested the PR in a Windows VM on VMware Fusion but I'd also appreciate any additional testing on any other Windows machine.

**Review by dbanck — COMMENTED** [review]
Thank you for answering all my questions and adding more tests!

Your refactorings around the handle/URI logic look solid, and the comments are helpful. I would never have guessed that VS Code is making paths so difficult.

Next, I'll test the changes on Windows.

File: internal/uri/uri.go [hunk 5]
```diff
@@ -26,24 +69,119 @@ func IsURIValid(uri string) bool {
...
+	})
 }
 
-func parseUri(uri string) (string, error) {
-	u, err := url.ParseRequestURI(uri)
+func MustPathFromURI(uri string) string {
+	osPath, err := PathFromURI(uri)
 	if err != nil {
-		return "", err
+		panic(fmt.Sprintf("invalid URI: %s", uri))
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply internal/uri/uri.go:128]: `panic` does shut it down, which is intended here in `Must*`. It is a convention in Go to call functions `Must` if they panic and provide single return value. See e.g. https://pkg.go.dev/regexp#MustCompile or https://pkg.go.dev/github.com/hashicorp/go-version#Must

The idea is that the `IsURIValid` is used in places where there is a higher likelihood of URI being invalid _and_ (crucially) there is a reasonable way of dealing with that situation more gracefully.

In most places like `initialize` or text-sync handlers I don't think we can do better than panic - we can't do much on `initialize` without a valid URI and we can't return error in LSP/RPC _notifications_ as notifications are not meant to be responded to by their nature.

In theory we could also error out in completion, hover etc. but I feel that we still _should_ be panicking there to prompt users to report this as a bug - either client-side or server-side. Also I'm not sure how clients actually deal with errors in such cases.

Custom commands will however error out.

[the remark this answers is not in the document]
Comment (dbanck) [reply internal/uri/uri.go:128]: Thanks for the detailed explanation! Calling `panic` in these situations looks like the right approach.

**Review by dbanck — APPROVED** [review]
🎉 🎉 🎉 

Everything works fine for me on Windows, too.
